### ISSUE TITLE
[Feature] Should we make the LM head GEMM output fp32 instead of bf16?

### ISSUE THREAD

**Issue by b8zhong** [sgl-project/sglang#33627]
### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Motivation

**Problem**

The LM head GEMM outputs bf16. We cast the result to fp32 afterwards, in `_copy_logits_to_buffer`. The cast happens after the rounding, so the fp32 precision the GEMM already computed internally is thrown away. It should strictly increase accuracy.

```python
# logits_processor.py::_compute_lm_head, default path
logits = torch.matmul(hidden_states.to(lm_head.weight.dtype), lm_head.weight.T)  # bf16 out
```

**Proposal**

Ask the GEMM for fp32 output instead:

```python
logits = torch.mm(hidden_states, lm_head.weight.T, out_dtype=torch.float32)
```

The weight stays bf16. Only the output dtype changes.

**Facts**

Measured on B300. DeepSeek-V4-Flash-0731 (vocab 129280, hidden 4096), TP4, so N=32320 per rank and K=4096. `bench_gpu_time_with_cupti` with CUDA graph.

1. Accuracy, against an fp32 reference. Current bf16 output changes the greedy token on 141/5120 positions (2.75%). `out_dtype=torch.float32` changes 0/5120.

2. Time. fp32 output makes the GEMM faster (it removes a separate cast kernel over the whole vocab) and the vocab all-gather slower (2x the bytes).

| M | GEMM now | GEMM fp32 | all-gather now | all-gather fp32 | net |
|---|---|---|---|---|---|
| 1 | 62.2us | 58.8us | 6.5us | 6.9us | -3.0us |
| 8 | 61.6us | 57.9us | 9.8us | 11.2us | -2.3us |
| 32 | 62.8us | 58.5us | 17.4us | 28.8us | +7.1us |
| 128 | 73.8us | 64.8us | 52.4us | 98.6us | +37.2us |
| 512 | 137.7us | 100.4us | 198.9us | 369.8us | +133.6us |

(all-gather column uses whichever of multimem / NCCL is faster at that size)

3. TP invariance. The LM head is column-parallel, so K is never sharded and every logit is the same dot product at any TP size. TP1 and TP4 agree exactly today, and would still agree if fp32 is applied everywhere. They would stop agreeing if fp32 is applied only when there is no all-gather.

| | greedy tokens that differ |
|---|---|
| now: TP1 bf16 vs TP4 bf16 | 0.00% |
| fp32 everywhere: TP1 vs TP4 | 0.00% |
| fp32 only when no all-gather: TP1 fp32 vs TP4 bf16 | 2.75% |

So this should be applied to all TP sizes or none.

**Open question**

Either flip the default to fp32 output, or leave the default and keep `--enable-fp32-lm-head` as the opt-in.

The overhead is likely negligible at large batch. The LM head runs once per forward, while this model has 43 transformer layers, and at large batch the decode step is dominated by those layers. So +133.6us at M=512 is probably a small fraction of the step. But we should benchmark it for TP case for some common models, like GLM-5.2, Deepseek V4, etc.

**Two unrelated things found while measuring**

- The multimem all-gather is slower than NCCL above M ~= 300 in bf16: 257.8us vs 198.9us at M=512, 1018us vs 705us at M=2048. `MultimemAllGatherer` guards on `recommended_max_tokens`, which is a token-count cap from `max_running_requests`, not a performance crossover.
- The multimem all-gather only supports bf16, so `--enable-fp32-lm-head` on TP>1 falls back to NCCL for the logits gather.

### Related resources

- #32370 uses `torch.mm(out_dtype=torch.float32)` for the same reason, on the `--enable-fp32-lm-head` path.
- `torch.mm` `out_dtype`: https://docs.pytorch.org/docs/2.13/generated/torch.mm.html

**Comment by b8zhong** [comment]
This issue is assigned
